pam/integration-tests/gdm-module-handler_test.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/ubuntu/authd"
+	"github.com/ubuntu/authd/internal/log"
+	"github.com/ubuntu/authd/pam/internal/gdm"
+	"github.com/ubuntu/authd/pam/internal/gdm_test"
+	"github.com/ubuntu/authd/pam/internal/proto"
+)
+
+type gdmTestModuleHandler struct {
+	t  *testing.T
+	tx *pam.Transaction
+
+	protoVersion uint32
+
+	supportedLayouts []*authd.UILayout
+
+	currentStage  proto.Stage
+	pollResponses []*gdm.EventData
+
+	authModes           []*authd.GAMResponse_AuthenticationMode
+	authModeID          string
+	selectedAuthModeIDs []string
+
+	brokersInfos       []*authd.ABResponse_BrokerInfo
+	brokerID           string
+	selectedBrokerName string
+
+	eventPollResponses map[gdm.EventType][]*gdm.EventData
+
+	pamInfoMessages  []string
+	pamErrorMessages []string
+}
+
+func (gh *gdmTestModuleHandler) exampleHandleGdmData(gdmData *gdm.Data) (*gdm.Data, error) {
+	switch gdmData.Type {
+	case gdm.DataType_hello:
+		return &gdm.Data{
+			Type:  gdm.DataType_hello,
+			Hello: &gdm.HelloData{Version: gh.protoVersion},
+		}, nil
+
+	case gdm.DataType_request:
+		return gh.exampleHandleAuthDRequest(gdmData)
+
+	case gdm.DataType_poll:
+		responses := gh.pollResponses
+		gh.pollResponses = nil
+		return &gdm.Data{
+			Type:         gdm.DataType_pollResponse,
+			PollResponse: responses,
+		}, nil
+
+	case gdm.DataType_event:
+		err := gh.exampleHandleEvent(gdmData.Event)
+		if err != nil {
+			return nil, err
+		}
+		return &gdm.Data{
+			Type: gdm.DataType_eventAck,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unhandled protocol message %s",
+		gdmData.Type.String())
+}
+
+func (gh *gdmTestModuleHandler) exampleHandleEvent(event *gdm.EventData) error {
+	events, ok := gh.eventPollResponses[event.Type]
+	if ok && len(events) > 0 {
+		pollResp := events[0]
+		gh.eventPollResponses[event.Type] = slices.Delete(events, 0, 1)
+		gh.pollResponses = append(gh.pollResponses, pollResp)
+	}
+
+	switch ev := event.Data.(type) {
+	case *gdm.EventData_BrokersReceived:
+		if len(ev.BrokersReceived.BrokersInfos) == 0 {
+			return errors.New("no brokers available")
+		}
+		gh.brokersInfos = ev.BrokersReceived.BrokersInfos
+
+		if gh.selectedBrokerName == ignoredBrokerName {
+			return nil
+		}
+
+		idx := slices.IndexFunc(gh.brokersInfos, func(bi *authd.ABResponse_BrokerInfo) bool {
+			return bi.Name == gh.selectedBrokerName
+		})
+		if idx < 0 {
+			return fmt.Errorf("broker '%s' is not known", gh.selectedBrokerName)
+		}
+
+		gh.pollResponses = append(gh.pollResponses, gdm_test.SelectBrokerEvent(gh.brokersInfos[idx].Id))
+
+	case *gdm.EventData_BrokerSelected:
+		idx := slices.IndexFunc(gh.brokersInfos, func(broker *authd.ABResponse_BrokerInfo) bool {
+			return broker.Id == ev.BrokerSelected.BrokerId
+		})
+		if idx < 0 {
+			return fmt.Errorf("unknown broker: %s", ev.BrokerSelected.BrokerId)
+		}
+		gh.brokerID = gh.brokersInfos[idx].Id
+		gh.t.Logf("Using broker '%s'", gh.brokersInfos[idx].Name)
+		require.Equal(gh.t, gh.selectedBrokerName, gh.brokersInfos[idx].Name)
```

Comment:
same here, please be more explicit with labelling the assertion so that we know what was the intent of the equality check. (same below)